File: OtterMath/Private/Otter/Math/Transform.c
#include "Otter/Math/Transform.h"

#include "Otter/Math/Mat.h"

void transform_identity(Transform* transform)
{
  transform->position.x = 0;
  transform->position.y = 0;
  transform->position.z = 0;

  transform->rotation.x = 0;
  transform->rotation.y = 0;
  transform->rotation.z = 0;

  transform->scale.x = 1;
  transform->scale.y = 1;
  transform->scale.z = 1;
}

void transform_apply(Mat4 matrix, Transform* transform)
{
  mat4_scale(
      matrix, transform->scale.x, transform->scale.y, transform->scale.z);
  mat4_rotate(matrix, transform->rotation.x, transform->rotation.y,
      transform->rotation.z);
  mat4_translate(matrix, transform->position.x, transform->position.y,
      transform->position.z);
}

File: GameScript/Source/Engine/Math/Transform.cs
namespace OtterEngine.Math
{
  using System;
  using System.Runtime.InteropServices;

  [StructLayout(LayoutKind.Sequential)]
  class Transform
  {
    public Vec3 position;
    public Vec3 rotation;
    public Vec3 scale;

    public Transform()
    {
      position = new Vec3(0, 0, 0);
      rotation = new Vec3(0, 0, 0);
      scale = new Vec3(1, 1, 1);
    }

    public Transform(Vec3 position, Vec3 rotation, Vec3 scale)
    {
      this.position = position;
      this.rotation = rotation;
      this.scale = scale;
    }
  }
}
